### ISSUE TITLE
Not adding Bot or BotVersion

### ISSUE THREAD

**Issue by marioconcilio** [scolladon/sfdx-git-delta#219]
Issue verification check :

- [x] does the current repository is fully deployable at the commit sha given to the from parameter of the command ?

## What is the problem

---

We are looking into using sgd to automate our deployments. During testing, we noticed that neither bots nor botVersions are being added to the package, although they have changed.

### What is the parameter and their value you used

```bash
sfdx sgd:source:delta \
    --from $(git merge-base issue/219 master) \
    --to HEAD \
    --output changed_sources \
    --api-version=53.0 \
    --source sgd \
```

### What is the expected result

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Package xmlns="http://soap.sforce.com/2006/04/metadata">
    <types>
        <members>Test</members>
        <members>Test.xml</members>
        <name>ApexClass</name>
    </types>
    <types>
        <members>BR_WhatsApp_Bot</members>
        <name>Bot</name>
    </types>
    <types>
        <members>BR_WhatsApp_Bot.v25</members>
        <name>BotVersion</name>
    </types>
    <version>53.0</version>
</Package>
```

### What is the actual result

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Package xmlns="http://soap.sforce.com/2006/04/metadata">
    <types>
        <members>Test</members>
        <members>Test.xml</members>
        <name>ApexClass</name>
    </types>
    <version>53.0</version>
</Package>
```

## Steps to reproduce

---

1. Branch off from master
2. Add a new bot and a botVersion to the sfdx project.
3. Commit the new `bot-meta.xml` and `botVersion-meta.xml` files
3. Run SGD with the following parameters (comparison copied from the 3rd topic of the [Diff between branches](https://github.com/scolladon/sfdx-git-delta#diff-between-branches) section:
```bash
--from $(git merge-base issue/219 master) # common ancestor
--to HEAD
--output changed_sources
--api-version=53.0
```

## Execution context

---

**Operating System:** MacOs Catalina (Darwin Kernel Version 19.4.0; root:xnu-6153.101.6~15/RELEASE_X86_640)

**node version:** v16.13.1

**git version:** git version 2.24.3 (Apple Git-128)

**sfdx version:** sfdx-cli/7.129.0 darwin-x64 node-v16.13.1

**sgd plugin version:** sfdx-git-delta 4.11.0

## Optional more information

---

`git diff --name-status --no-renames $(git merge-base issue/219 master) HEAD`

Output:
```
A       sgd/reproduction/playground/bots/BR_WhatsApp_Bot/BR_WhatsApp_Bot.bot-meta.xml
A       sgd/reproduction/playground/bots/BR_WhatsApp_Bot/v25.botVersion-meta.xml
A       sgd/reproduction/playground/classes/Test.cls
A       sgd/reproduction/playground/classes/Test.xml
```

**Comment by marioconcilio** [comment]
I tried to push my example to https://github.com/scolladon/sfdx-git-delta-reproduction-playground, but I have no access. 
BTW, that is a great plugin, nice job!

**Comment by scolladon** [comment]
Hi @marioconcilio !

Thanks for this very well documented issue !
Thanks for your help to make this plugin better !

You should now have access to the reproduction playground, if you could create a branch there (`issues/219` per example) and put just what it needs to have to isolate the issue it would be great.

I'll look at it soon and let you know here my findings !

Thanks for your compliments about the plugin, it is always very nice to hear that, thank you from the whole team

**Comment by marioconcilio** [comment]
Hi @scolladon! Thanks for the prompt response!
I cloned the repo and created the branch, but I still don't have permission to push my changes. Can you double-check it, please?

**Comment by scolladon** [comment]
You should have receive an email to accept the access to the repo
here is the [link](https://github.com/scolladon/sfdx-git-delta-reproduction-playground/invitations)

**Comment by marioconcilio** [comment]
Oops, my mistake! I've just pushed the new branch. I am at your disposal to test the new solution and clarify any doubts. 

Thanks again for your attention!

**Comment by scolladon** [comment]
Thanks for providing data to isolate the behaviour

I am trying to compare what is in the [reproduction-playground](https://github.com/scolladon/sfdx-git-delta-reproduction-playground/tree/issue/219) with what the [doc says](https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_botversion.htm)

Bot and BotVersion are already in our [metadata file](https://github.com/scolladon/sfdx-git-delta/blob/fe7ace3f3e3b6f3b39408fc19ec7f41c96245cfd/src/metadata/v52.json#L1208) so they should be handled (and [they are](https://github.com/scolladon/sfdx-git-delta/pull/148) but not with this form).

What I can see with this issue and the reproduction playground is that it is expected Bot and BotVersion to be retrieved each in their own files. 
What looks weird to me is how to deduce to which Bot the BotVersion is related to? 
Because the name contains only the version number and not the Bot to which it is related (the file does not contains any reference either). Then the name would be taken from the subfolder where it is located I guess ?

But when I look at the doc I can see only one file (`.bot`) containing the [`botVersion` definition](https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_botversion.htm#:~:text=Astros_Pizza_ld1%3C/name%3E%0A%20%20%20%20%3C/botMlDomain%3E-,%3CbotVersions%3E,-%3CfullName%3Ev1%3C/fullName).

But ! the [Bot suffix and Directory location](https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_botversion.htm#:~:text=File%20Suffix%20and%20Directory%20Location) section can also be understood as if there is 2 different files, one of the bot definition and one for the bot version, with the same file extension... (weird)

@marioconcilio if you have any experience with that, answers to those questions, documentations, please share with us.
And also, how did you to get those files ? with which api version ? Does the result is the same on every API version which support Bot and BotVersion metadata ?

**Comment by mehdisfdc** [comment]
@scolladon I've performed a few tests, and I confirm that a bot is actually composed of 2 files (the `.bot-meta.xml` and the `.botVersion-meta.xml`). I believe the doc is misleading regarding the suffix of the BotVersion.
Both are located **within the same sub-folder** (named as the bot) inside the `bot` folder.

For example, I've created 3 bots in a scratch org, and here is the retrieved metadata after running a `sfdx force:source:pull`:
- 3 subfolders (1 per bot) are created within the bot folder:
![bots](https://user-images.githubusercontent.com/25982646/145555769-419b7923-9cbb-487b-be83-bba27f60b875.png)
- each subfolder is composed of 2 files (the `.bot-meta.xml` and the `.botVersion-meta.xml` for the bot)
![botFiles](https://user-images.githubusercontent.com/25982646/145555911-d5ef1f98-1a26-45b1-8250-063e5b003f48.png)

**Comment by marioconcilio** [comment]
Hi @scolladon, sorry for the late response.

> What looks weird to me is how to deduce to which Bot the BotVersion is related to?
Because the name contains only the version number and not the Bot to which it is related (the file does not contains any reference either). Then the name would be taken from the subfolder where it is located I guess ?

The botVersion is related to the Bot that is on the same subfolder. It is just as @mehdisfdc stated. The documentation is indeed misleading regarding the bot suffix.

> And also, how did you to get those files ? with which api version ? Does the result is the same on every API version which support Bot and BotVersion metadata ?

Bots are retrieved using the following command:
`sfdx force:source:retrieve -m Bot:BR_SME_WhatsApp_Bot`

This retrieves the `bot-meta.xml` file and all its version files (`botVersion-meta.xml`):
```bash
FULL NAME            TYPE        PROJECT PATH
───────────────────  ──────────  ────────────────────────────────────────────────────────────────────────────
BR_WhatsApp_Bot      Bot         force-app/main/default/bots/BR_SME_WhatsApp_Bot/BR_WhatsApp_Bot.bot-meta.xml
v25                  BotVersion  force-app/main/default/bots/BR_SME_WhatsApp_Bot/v25.botVersion-meta.xml
```

Additionally, it is possible to retrieve a single bot version running:
`sfdx force:source:retrieve -m BotVersion:BR_WhatsApp_Bot.v25`

This retrieves a single file representing the bot version (`botVersion-meta.xml`):
```bash
FULL NAME  TYPE        PROJECT PATH
─────────  ──────────  ───────────────────────────────────────────────────────────────────
v25        BotVersion  force-app/main/default/bots/BR_WhatsApp_Bot/v25.botVersion-meta.xml
```

**Comment by scolladon** [comment]
Hi @marioconcilio !

I just created a new [version](https://github.com/scolladon/sfdx-git-delta/pull/226) (currently under review).

Could you try it (how to [here](https://github.com/scolladon/sfdx-git-delta/blob/master/CONTRIBUTING.md#testing-the-plugin-from-a-pull-request)) and give feedback ?

**Comment by marioconcilio** [comment]
Hi @scolladon! 

I tested and it worked! The plugin added the `bot` and `botVersion` to the changed sources. 
There is just one detail - and I forgot to mention: `botVersions` cannot be deployed alone, they need the `bot` file as well, just like it happens with `fieldTranslations` (they need the `objectTranslation` too, and the sgd plugin automatically includes them).

Can you please make sgd automatically include `bot` files whenever a `botVersion` is added to the package?

**Comment by scolladon** [comment]
Hi @marioconcilio !

I just added this behaviour in the PR #226 
Ready for you to check how it works.
I added a commit in the [reproduction-playground issue/219](https://github.com/scolladon/sfdx-git-delta-reproduction-playground/tree/issue/219) branch to isolate the behaviour

Let me know !

**Comment by marioconcilio** [comment]
Hi @scolladon!

I just tested and the `bot-meta.xml` files are being copied to the right place when running with the `--generate-delta` flag. However, the `package.xml` is wrong when we only add/change a botVersion (`botVersion-meta.xml` files)

```xml
<types>
    <members>BRAcquisitionWhatsAppBot</members>
    <members>BR_WhatsApp_Bot</members>
    <members>Sam_Web_IT</members>
    <name>bots</name>
</types>
<types>
    <members>BRAcquisitionWhatsAppBot.v5</members>
    <members>BR_WhatsApp_Bot.v26</members>
    <members>Sam_Web_IT.v12</members>
    <name>BotVersion</name>
</types>
```

The expected output is:
```xml
<types>
    <members>BRAcquisitionWhatsAppBot</members>
    <members>BR_WhatsApp_Bot</members>
    <members>Sam_Web_IT</members>
    <name>Bot</name>
</types>
<types>
    <members>BRAcquisitionWhatsAppBot.v5</members>
    <members>BR_WhatsApp_Bot.v26</members>
    <members>Sam_Web_IT.v12</members>
    <name>BotVersion</name>
</types>
```

When we change only the bot (`bot-meta.xml` files), the package is generated accordingly:
```xml
<types>
    <members>BR_WhatsApp_Bot</members>
    <name>Bot</name>
</types>
```

Please let me know when it is ready to test!

**Comment by marioconcilio** [comment]
It is working perfectly fine @scolladon !

I tested creating a new bot, and the sgd:
- copied the `bot` and `botVersion` files to the changed sources folder accordingly
- created a valid `package.xml` with `Bot` and `BotVersion` types

I also tested creating just a new bot version, and the sgd:
- copied the `bot` and the new `botVersion` files to the changed sources folder accordingly
- created a valid `package.xml` with `Bot` and the new `BotVersion` types

Additionally, I tested deleting one single bot version, and the sgd:
- created a valid `destructiveChanges.xml` with the `BotVersion` type to delete

Finally, I tested deleting the entire bot, and the sgd:
- created a valid `destructiveChanges.xml` with the `Bot` and the `BotVersion` to delete

In summary, **sfdx-git-delta** is working:
- to create a new Bot ✅ 
- to create a new BotVersion ✅ 
- to delete a single BotVersion ✅ 
- to delete an entire Bot (w/ its BotVersions) ✅ 

Thanks a lot for this huge fix! I am looking forward to seeing this new version published!

**Comment by scolladon** [comment]
Thank you very much for the QA and all the help !

**Comment by mehdisfdc** [comment]
Thank you @scolladon for the code, and thank you @marioconcilio for find the issue and testing the new version! The plugin is getting better every week 😄

**Comment by scolladon** [comment]
New version ([4.12.0](https://www.npmjs.com/package/sfdx-git-delta)) available !
